sway/input/cursor.c

```diff
@@ -242,8 +464,79 @@ static void handle_cursor_motion_absolute(
 	cursor_send_pointer_motion(cursor, event->time_msec, true);
 }
 
+static void handle_end_operation(struct sway_seat *seat) {
+	if (seat->operation == OP_DRAG) {
+		// We "move" the container to its own location so it discovers its
+		// output again.
+		struct sway_container *con = seat->op_container;
+		container_floating_move_to(con, con->x, con->y);
+		seat->operation = OP_NONE;
+		seat->op_container = NULL;
+	} else {
+		// OP_RESIZE
+		seat->operation = OP_NONE;
+		seat->op_container = NULL;
+	}
+}
+
+static void dispatch_cursor_button_floating(struct sway_cursor *cursor,
+		uint32_t time_msec, uint32_t button, enum wlr_button_state state,
+		struct wlr_surface *surface, double sx, double sy,
+		struct sway_container *cont) {
+	struct sway_seat *seat = cursor->seat;
+
+	// Deny dragging or resizing a fullscreen view
+	if (cont->type == C_VIEW && cont->sway_view->is_fullscreen) {
+		wlr_seat_pointer_notify_button(seat->wlr_seat, time_msec, button, state);
+		return;
+	}
+
+	struct wlr_keyboard *keyboard = wlr_seat_get_keyboard(seat->wlr_seat);
+	bool mod_pressed = keyboard &&
+		(keyboard->modifiers.depressed & config->floating_mod);
+	enum resize_edge edge = find_resize_edge(cont, cursor);
+	bool over_title = edge == RESIZE_EDGE_NONE && !surface;
+
+	// Check for beginning drag
+	if (button == BTN_LEFT && state == WLR_BUTTON_PRESSED &&
+			(mod_pressed || over_title)) {
+		seat->operation = OP_DRAG;
+		seat->op_container = cont;
+		seat->op_button = button;
+		return;
+	}
+
+	// Check for beginning resize
+	bool resizing_via_border = button == BTN_LEFT && edge;
+	bool resizing_via_mod = button == BTN_RIGHT && mod_pressed;
+	if ((resizing_via_border || resizing_via_mod) &&
+			state == WLR_BUTTON_PRESSED) {
+		seat->operation = OP_RESIZE;
+		seat->op_container = cont;
+		seat->op_resize_preserve_ratio = keyboard &&
+			(keyboard->modifiers.depressed & 1); // Shift
```

Comment:
If anyone knows how I can replace `1` with a constant or more meaningful value, let me know.

Reply:
WLR_MODIFIER_SHIFT is defined in wlroots/include/wlr/types/wlr_keyboard.h. 

Is using the structure directly fine here or should you call wlr_keyboard_get_modifiers?

Reply:
I replaced the value - thanks.

I'm not sure what `wlr_keyboard_get_modifiers` is doing, so I'll wait for someone more knowledgeable to let me know if I should use it and what the consequences are of using the struct directly.

Reply:
`WLR_MODIFIER_*` constants cannot be used with `keyboard->modifiers.depressed`, they can only be used with `wlr_keyboard_get_modifiers`, because modifier indexes are not guaranteed to be the same from one keyboard layout to another.